crates/clvm-utils/src/tree_hash.rs

```diff
@@ -307,6 +308,79 @@ pub fn tree_hash_cached(a: &Allocator, node: NodePtr, cache: &mut TreeCache) ->
     hashes[0]
 }
 
+fn subtract_cost(cost_left: &mut Cost, subtract: Cost) -> Result<(), usize> {
+    if subtract > *cost_left {
+        Err(0)
+    } else {
+        *cost_left -= subtract;
+        Ok(())
+    }
+}
+
+pub fn tree_hash_cached_costed(
+    a: &Allocator,
+    node: NodePtr,
+    cache: &mut TreeCache,
+    cost_left: &mut u64,
+    cost_per_call: u32,
+    cost_per_byte: u32,
+) -> Result<TreeHash, usize> {
+    cache.visit_tree(a, node);
+
+    let mut hashes = Vec::new();
+    let mut ops = vec![TreeOp::SExp(node)];
+
+    while let Some(op) = ops.pop() {
+        subtract_cost(cost_left, cost_per_call.into())?;
+        match op {
+            TreeOp::SExp(node) => match a.node(node) {
+                NodeVisitor::Buffer(bytes) => {
+                    subtract_cost(cost_left, (cost_per_byte * bytes.len() as u32).into())?;
+                    let hash = tree_hash_atom(bytes);
+                    hashes.push(hash);
+                }
+                NodeVisitor::U32(val) => {
+                    subtract_cost(cost_left, (cost_per_byte * 4).into())?; // 4 bytes in u32
+                    if (val as usize) < PRECOMPUTED_HASHES.len() {
+                        hashes.push(PRECOMPUTED_HASHES[val as usize]);
+                    } else {
+                        hashes.push(tree_hash_atom(a.atom(node).as_ref()));
+                    }
+                }
+                NodeVisitor::Pair(left, right) => {
+                    subtract_cost(cost_left, cost_per_byte.into())?;
+                    if let Some(hash) = cache.get(node) {
+                        hashes.push(*hash);
```

Comment:
this case is not accounting cost correctly. You should have a test case for this, that ensures the same tree always has the same cost to tree-hash, regardless of cache hits.